sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/plugin/caching/ContentHashCache.kt

```diff
@@ -0,0 +1,114 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.plugin.caching
+
+import org.sonar.api.batch.fs.InputFile
+import org.sonar.api.batch.sensor.SensorContext
+import org.sonar.api.batch.sensor.cache.ReadCache
+import org.sonar.api.batch.sensor.cache.WriteCache
+import org.sonar.api.utils.log.Loggers
+import java.security.MessageDigest
+
+private val LOG = Loggers.get(ContentHashCache::class.java)
+private const val HASH_ALGORITHM = "MD5"
+private val messageDigest = MessageDigest.getInstance(HASH_ALGORITHM)
+const val CONTENT_HASHES_KEY = "kotlin:contentHash:$HASH_ALGORITHM:"
+
+class ContentHashCache {
+
+    private var enabled = false
+    private var readCache: ReadCache? = null
+    private var writeCache: WriteCache? = null
+    private var computedHashes = mutableMapOf<String, ByteArray>()
+
+    constructor(sensorContext: SensorContext){
+        this.enabled = sensorContext.isCacheEnabled
+        if(this.enabled){
+            LOG.info("Content Hash Cache was initialized")
+            this.readCache = sensorContext.previousCache()
+            this.writeCache = sensorContext.nextCache()
+        }
+    }
+
+    fun isEnabled() = this.enabled
+
+    fun hasDifferentContentCached(inputFile: InputFile): Boolean{
+        val key = contentHashKey(inputFile)
+        val cachedFileHash = read(key) ?: return true
+        val inputFileHash = getHash(inputFile)
+        computedHashes[key] = inputFileHash
+        return !MessageDigest.isEqual(inputFileHash, cachedFileHash)
+    }
+
+    fun writeContentHash(inputFile: InputFile){
+        val key = contentHashKey(inputFile)
+        if(computedHashes.contains(key)){
+            LOG.trace("Using already processed hash for key: $key")
+            write(key, computedHashes.remove(key)!!)
+        }else{
+            write(key, getHash(inputFile))
+        }
+    }
+
+    fun copyFromPreviousAnalysis(inputFile: InputFile){
+        if(isEnabled()){
+            val key = contentHashKey(inputFile)
+            if(readCache!!.contains(key)){
+                writeCache!!.copyFromPrevious(key)
+                LOG.trace("Copied file $key from previous cache.")
+            }else{
+                writeContentHash(inputFile)
+            }
+        }else{
+            LOG.trace("Cannot write to cache when disabled.")
+        }
+    }
+
+    private fun write(key: String, hash: ByteArray){
+        if(isEnabled()){
+            LOG.trace("Storing file content hash for key: $key")
+            writeCache!!.write(key, hash)
+        }else{
+            LOG.trace("Cannot write to cache when disabled.")
+        }
+    }
+
+    private fun read(key: String): ByteArray? {
+        if(enabled){
+            if(readCache!!.contains(key)){
+                val stream = readCache!!.read(key)
+                return stream.readAllBytes().also { stream.close() }
```

Comment:
Let's use "try-with-resources" here. In Kotlin, this is done via the `.use` method.